File: android/Book-AndroidProgramming/CriminalIntentJ/app/src/main/java/kr/ac/ajou/heidi/criminalintentj/controller/CrimeListFragment.java
package kr.ac.ajou.heidi.criminalintentj.controller;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;
import java.util.UUID;

import kr.ac.ajou.heidi.criminalintentj.R;
import kr.ac.ajou.heidi.criminalintentj.model.Crime;
import kr.ac.ajou.heidi.criminalintentj.model.CrimeLab;

public class CrimeListFragment extends Fragment {
    private RecyclerView mCrimeRecyclerView;
    private CrimeAdapter mAdapter;
    private int resultCrimePosition;

    private class CrimeHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView mTitleTextView;
        private TextView mDateTextView;
        private CheckBox mSolvedCheckBox;
        private Crime mCrime;

        public CrimeHolder(@NonNull View itemView) {
            super(itemView);

            mTitleTextView = itemView.findViewById(R.id.list_item_crime_title_text_view);
            mDateTextView = itemView.findViewById(R.id.list_item_crime_date_text_view);
            mSolvedCheckBox = itemView.findViewById(R.id.list_item_crime_solved_check_box);

            itemView.setOnClickListener(this);
        }

        public void bindCrime(Crime crime) {
            mCrime = crime;
            mTitleTextView.setText(mCrime.getmTitle());
            mDateTextView.setText(mCrime.getmDate());
            mSolvedCheckBox.setChecked(mCrime.ismSolved());
        }

        @Override
        public void onClick(View v) {
//            Intent intent = CrimeActivity.newIntent(getActivity(), mCrime.getmId());
            Intent intent = CrimePagerActivity.newIntent(getActivity(), mCrime.getmId());
            startActivityForResult(intent, 123);
        }
    }

    private class CrimeAdapter extends RecyclerView.Adapter<CrimeHolder> {

        private List<Crime> mCrimes;

        public CrimeAdapter(List<Crime> crimes) {
            mCrimes = crimes;
        }

        @NonNull
        @Override
        public CrimeHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            View view = layoutInflater.inflate(R.layout.list_item_crime, parent, false);
            return new CrimeHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull CrimeHolder crimeHolder, int position) {
            Crime crime = mCrimes.get(position);
            crimeHolder.bindCrime(crime);
        }

        @Override
        public int getItemCount() {
            return mCrimes.size();
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_crime_list, container, false);
        mCrimeRecyclerView = view.findViewById(R.id.crime_recycler_view);
        mCrimeRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        updateUI();

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        updateUI();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            UUID crimeId = (UUID) data.getSerializableExtra("crime_id");
            for (int i = 0; i < mAdapter.mCrimes.size(); i++) {
                if (mAdapter.mCrimes.get(i).getmId() == crimeId) {
                    resultCrimePosition = i;
                }
            }
        }
    }

    private void updateUI() {
        CrimeLab crimeLab = CrimeLab.get(getActivity());
        List<Crime> crimes = crimeLab.getCrimes();

        if (mAdapter == null) {
            mAdapter = new CrimeAdapter(crimes);
            mCrimeRecyclerView.setAdapter(mAdapter);
        } else {
//            mAdapter.notifyDataSetChanged();
            mAdapter.notifyItemChanged(resultCrimePosition);
        }

    }
}

File: android/Book-AndroidProgramming/CriminalIntentK/app/src/main/java/kr/ac/ajou/heidi/criminalintentk/controller/CrimeListFragment.kt
package kr.ac.ajou.heidi.criminalintentk.controller

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.*
import kotlinx.android.synthetic.main.fragment_crime_list.view.*
import kotlinx.android.synthetic.main.list_item_crime.view.*
import kr.ac.ajou.heidi.criminalintentk.R
import kr.ac.ajou.heidi.criminalintentk.model.Crime
import kr.ac.ajou.heidi.criminalintentk.model.CrimeLab

class CrimeListFragment : Fragment() {

    inner class CrimeViewHolder(parent: ViewGroup) :
            RecyclerView.ViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.list_item_crime, parent, false)), View.OnClickListener {


        lateinit var crime: Crime

        init {
            this.itemView.setOnClickListener(this)

        }

        fun setCrimeObject(crime: Crime) {
            this.crime = crime
        }


        override fun onClick(view: View?) {

            val intent = CrimePagerActivity.newIntent(activity!!.applicationContext, crime.id)
            startActivityForResult(intent, 123)
        }

    }


    inner class CrimeViewAdapter : RecyclerView.Adapter<CrimeViewHolder>() {
        var crimes: ArrayList<Crime> = arrayListOf()

        override fun onCreateViewHolder(parent: ViewGroup, position: Int)
                : CrimeViewHolder = CrimeViewHolder(parent)

        override fun getItemCount(): Int = crimes.size

        override fun onBindViewHolder(crimeHolder: CrimeViewHolder, position: Int) {
            val crime = crimes[position]

            crimeHolder.setCrimeObject(crime)

            with(crimeHolder.itemView) {
                crimeTitleTextView.text = crime.title
                crimeDateTextView.text = crime.date.toString()
                crimeSolvedCheckBox.isChecked = crime.solved
            }
        }
    }

    companion object {
        const val SAVED_SUBTITLE_VISIBLE = "subtitle"
    }

    private val adapter = CrimeViewAdapter()
    private var subtitleVisible: Boolean = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        savedInstanceState?.let {
            subtitleVisible = it.getBoolean(SAVED_SUBTITLE_VISIBLE)
        }

        val view = inflater.inflate(R.layout.fragment_crime_list, container, false)
        view.crimeRecyclerView.layoutManager = LinearLayoutManager(activity!!.applicationContext)
        val crimeLab = CrimeLab.get(view?.context ?: return view)
        view.crimeRecyclerView.adapter = adapter
        adapter.crimes = crimeLab?.crimes ?: return view

        return view
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.menu_item_new_crime -> {
                val crime = Crime()
                CrimeLab.get(view?.context ?: return false)?.addCrime(crime)
                view?.let {
                    val intent = CrimePagerActivity.newIntent(it.context, crime.id)
                    startActivity(intent)
                }
                return true
            }

            R.id.menu_item_show_subtitle -> {
                subtitleVisible = !subtitleVisible
                activity?.invalidateOptionsMenu()
                updateSubtitle()
                return true
            }

            else -> {
                super.onOptionsItemSelected(item)
                return false
            }
        }

    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater?.inflate(R.menu.fragment_crime_list, menu)

        val subtitleItem = menu?.findItem(R.id.menu_item_show_subtitle) ?: return
        if (subtitleVisible) {
            subtitleItem.setTitle(R.string.hide_subtitle)
        } else {
            subtitleItem.setTitle(R.string.show_subtitle)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 123) {

        }
    }

    override fun onResume() {
        super.onResume()
        updateUI()

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    private fun updateUI() {

        CrimeLab.get(view?.context ?: return)?.let {
            adapter.crimes = it.crimes

        }
        adapter.notifyDataSetChanged()

        updateSubtitle()
    }

    private fun updateSubtitle() {
        val crimeLab = CrimeLab.get(view?.context ?: return)
        val crimeCount = crimeLab?.crimes?.size ?: 0
        var subtitle: String? = resources.getQuantityString(
                R.plurals.subtitle_plural, crimeCount, crimeCount
        )

        if (!subtitleVisible) {
            subtitle = null
        }

        val activity = activity as AppCompatActivity
        activity.supportActionBar?.subtitle = subtitle

    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(SAVED_SUBTITLE_VISIBLE, subtitleVisible)
    }
}